descartes_light/core/include/descartes_light/core/edge_evaluator.h

```diff
@@ -35,7 +35,12 @@ class EdgeEvaluator
   using Ptr = typename std::shared_ptr<EdgeEvaluator<FloatType>>;
   using ConstPtr = typename std::shared_ptr<const EdgeEvaluator<FloatType>>;
 
+  EdgeEvaluator() = default;
   virtual ~EdgeEvaluator() = default;
+  EdgeEvaluator(const EdgeEvaluator&) = default;
+  EdgeEvaluator& operator=(const EdgeEvaluator&) = default;
+  EdgeEvaluator(EdgeEvaluator&&) noexcept = default;
+  EdgeEvaluator& operator=(EdgeEvaluator&&) noexcept = default;
```

Comment:
Should abstract interfaces have valid constructors and copy/move assigmnent operators? I would think they should equal `delete` instead of `default`

Reply:
I recently ran into this. If it is not an abstract then you should use delete to prevent slicing, but if it is abstract it already cannot be copied and by making this default you would not be imposing this restriction on the derived class. This way derived class can have a copy constructor. I did this all over tesseract and I have slowly been fixing this for abstract classes. Now if you inherit from a derived class and store it as the intermediate class then copy it would slice but I believe this is very unlikely to happen.

Reply:
@marip8 @Levi-Armstrong should we be preferring to use the [AllowSoleDefaultDtor](https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines-special-member-functions.html#cmdoption-arg-allowsoledefaultdtor) clang-tidy option here instead? I think these additions are just unnecessary noise, no?

Reply:
I think that would be fine. I was not aware of the option, good find. If @marip8 also agrees we can update the checks.

Reply:
My understanding is that:
1. Per [ISO c++](https://isocpp.org/wiki/faq/virtual-functions#virtual-dtors), base classes need to define a virtual destructor if classes derived from the base class will ever be deleted later
2. Per [ISO c++](https://isocpp.org/wiki/faq/virtual-functions#virtual-dtor-rationale), destructors are not virtual by default, so they need to be explicitly added

It seems like removing the explicit default virtual destructors might cause the compiler to define default non-virtual destructors which could cause issues later with the derived classes

Reply:
The destructor is fine, the issue is that its presence triggers clang-tidy to suggest the other operations must be added due to the rule of five. But a default destructor isn't changing the default behavior and so adding these other explicit definitions is noisy and has no benefit. Hence they've added the option linked above that allows this situation to go unmodified.